src/features/atomics.md

```diff
@@ -0,0 +1,202 @@
+
+Starting from Chrome 87, V8 supports a non-blocking version, [Atomics.waitAsync](https://github.com/tc39/proposal-atomics-wait-async/blob/master/PROPOSAL.md), which is also usable on the main thread.
+
+In this post, we explain how to use these low-level APIs to implement a mutex that works both synchronously (for worker threads) and asynchronously (for worker threads or the main thread).
+
+`Atomics.wait` and `Atomics.waitAsync` take the following parameters:
+
+- an `Int32Array` or `BigInt64Array` backed by a `SharedArrayBuffer`
+- index (valid within the array)
+- a value we expect to be present in the memory location pointed to by buffer and index
+- timeout in milliseconds (optional, defaults to `Infinity`)
+
+The return value of `Atomics.wait` is a string. If the memory location doesn’t contain the expected value, `Atomics.wait` returns immediately with the value `"not-equal"`. Otherwise, the thread is blocked until another thread calls `Atomics.notify` with the same memory location or the timeout is reached. In the former case, `Atomics.wait` returns the value `"ok"`, in the latter case, `Atomics.wait` returns the value `"timed-out"`.
+
+`Atomics.notify` takes the following parameters:
+
+- an `Int32Array` or `BigInt64Array` backed by a `SharedArrayBuffer`
+- index (valid within the array)
+- how many waiters to notify (optional, defaults to `Infinity`)
+
+It notifies the given amout of waiters, in FIFO order, waiting on the memory location described by (buffer, index). If there are several pending `Atomics.wait` calls or `Atomics.waitAsync` calls related to the same location, they are all in the same FIFO queue.
+
+In contrast to `Atomics.wait`, `Atomics.waitAsync` always returns immediately. The return value is one of the following:
+
+- `{ async: false, value: "not-equal" }` (if the memory location didn’t contain the expected value)
+- `{ async: false, value: "timed-out" }` (only for immediate timeout 0)
+- `{ async: true, value: promise }`
+
+The promise may later be resolved with a string value `"ok"` (if `Atomics.notify` was called with the same memory location) or `"timed-out"` (if the timeout was reached). The promise is never rejected.
+
+The following example demonstrates the basic usage of `Atomics.waitAsync`:
+
+```js
+let sab = new SharedArrayBuffer(16);
+let i32a = new Int32Array(sab);
+let result = Atomics.waitAsync(i32a, 0, 0, 1000);
+//                                         ^ timeout (opt)
+//                                      ^ expected value
+//                                   ^ index
+
+if (result.value === "not-equal") {
+  // The value in the SharedArrayBuffer was not the expected one.
+} else {
+  result.value instanceof Promise; // true
+  result.value.then(
+    (value) => { if (value == "ok") { /* notified */ }
```

Comment:
Formatting is a bit odd here, let's split into a new line.

Reply:
done